### ISSUE TITLE
Chapter 14: Over 92% test accuracy on Fashion MNST?

### ISSUE THREAD

**Issue by hakank** [ageron/handson-ml2#66]
Thanks for a great book.

I'm preparing an assignment for a DL course where the objective is to get at least 90% test accuracy on Fashion MNIST using a CNN of our own.  

On page 461 the book show a Fashion MNIST model with the following comment (page 462):

_This CNN reaches over 92% accuracy on the test set. It's not state of the art, but it is pretty good,
and clearly much better than what we achieved with dense networks in Chapter 10._

This sounds interesting so I planned to use that network as a baseline to beat - or at least compare - with my quite smaller CNN (which almost always reach a little over 92% test accuracy).  You don't write the specifics of the training parameters, so I wonder how you get 92% test accuracy. 

When I run the network using Adam/NAdam and which early stopping (`patience=20`), I only get about 90% and tops 91.4% test accuracy using tf.keras with TF 2.

The training/test results that's shown in the current version of the chapter notebook (https://github.com/ageron/handson-ml2/blob/master/14_deep_computer_vision_with_cnns.ipynb ) just get 0.8983 test accuracy. However, it just run for exactly 10 epochs so it's probably not the proper model.

Would you mind showing the CNN model + the training parameters that give over 92% test accuracy?

Best regards,

Hakan

**Comment by klyuhang9** [comment]
I think there is no problem, recently I also done this assignment. And I did get 92.04% test accuracy on this CNN network on books at some epoches(The final accuracy will fluctuate a little)

**Comment by hakank** [comment]
Thanks for your comment. 

Did you run the training as it was in the book's notebook to get > 92%, or did you used some other training setup?

**Comment by ageron** [comment]
Hi @hakank,

Thanks for your question! Perfect reproducibility is tricky when dealing with Neural Networks (and many other ML models). First, the training algorithm (stochastic gradient descent) relies on randomness. This makes reproducibility harder, but not impossible: we can theoretically just set the random seed. Unfortunately the pseudo-random number generators may vary across version or even platforms. In particular, the logic changed between TF 1.x and TF 2.0. In fact, it changed as I was writing the second edition, which means that you may get slightly different results. They should be in the same ballpark, though.

For example, in TF 2.0:

```python
import tensorflow as tf
tf.random.set_seed(42)
print(tf.random.uniform([5]))
# outputs [0.6645621  0.44100678 0.3528825  0.46448255 0.03366041]
```

And in TF 1.15:

```python
import tensorflow as tf
tf.set_random_seed(42)
rnd = tf.random_uniform([5])
with tf.Session() as sess:
    print(rnd.eval())
# outputs [0.63789964 0.16047871 0.19355309 0.25280082 0.31931257]
```

Moreover, the random number logic is going to change again soon, see [this issue](https://github.com/tensorflow/tensorflow/issues/33297). Hopefully, it will remain stable for a long time after that! :)

But TensorFlow's random seeds are not the only difficulty. You need to set NumPy's random seed, Python's random seed, and even the random seed used by Python to generate hashes for sets and dicts. You also need to be careful with multithreading, in particular when using a GPU. I did a [video on the topic of reproducibility](https://www.youtube.com/watch?v=Ys8ofBeR2kA), in case you're interested.

In short, perfect reproducibility is hard to achieve and comes with severe constraints, so it might be preferable to tolerate some variability. If you're writing tests (or a book or a blog post) where you really absolutely need perfect reproducibility, then watch my video. Or just run the program 10 times and compute the mean, and check that it falls within a reasonable range.

I hope this helps!

**Comment by hakank** [comment]
Hi @ageron .

Thanks, your answer helps somewhat. I'm quite aware about the problem regarding reproducibility and how unreliable it is to just use the values of a single run. You video about reproducibility in TensorFlow was interesting - I didn't realize that it was _that_ hard to get deterministic results in TF - but I was not really after that precise reproduction of the test accuracy. 

My main question was more about the details how you trained the model, and I would then see what result I would get (given that we probably use different versions of TF, Python etc).

Here's how I trained your network, using early stopping, validation_split which surely influences the exact results (and randomness):
```
geron_model = keras.models.Sequential([
    keras.layers.Conv2D(64, 7, activation="relu", padding="same",
                        input_shape=[28, 28, 1]),
    keras.layers.MaxPooling2D(2),
    keras.layers.Conv2D(128, 3, activation="relu", padding="same"),
    keras.layers.Conv2D(128, 3, activation="relu", padding="same"),
    keras.layers.MaxPooling2D(2),
    keras.layers.Conv2D(256, 3, activation="relu", padding="same"),
    keras.layers.Conv2D(256, 3, activation="relu", padding="same"),
    keras.layers.MaxPooling2D(2),
    keras.layers.Flatten(),
    keras.layers.Dense(128, activation="relu"),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(64, activation="relu"),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(10, activation="softmax")
])

geron_model.compile(optimizer='nadam',
                       loss='sparse_categorical_crossentropy',
                       metrics=['accuracy'])

early_stopping_cb = keras.callbacks.EarlyStopping(monitor="val_loss", patience=20, restore_best_weights=True)

geron_history = geron_model.fit(train_images,
                                train_labels,
                                epochs=200,
                                validation_split=0.2,
                                callbacks=[early_stopping_cb]
                                )

geron_test_loss, geron_test_acc = geron_model.evaluate(test_images,  test_labels, verbose=2)
```

In TF 2 I ran the model a couple of times and got the following test accuracy:

      0.9082 0.9144 0.9113 0.9098 0.9058 0.9096 0.9149 0.9091 0.9103 0.9070 0.9075

with an average of 0.91. Testing with TF 1.15 (Colab's default TF) instead, didn't change much: 

   0.9105 0.9046 0.9084

with an average of 0.907, i.e. about the same.

Best regards,

Hakan

**Comment by ageron** [comment]
The code I ran is the code in the notebook, section "Tackling Fashion MNIST With a CNN". Indeed, the last times I ran it, the performance did not reach 92%, so perhaps I got a bit lucky the first time I ran it. It's more like 90.5% to 91.5%, in general.

**Comment by hakank** [comment]
@ageron Thanks, that answered my question.
